### PR TITLE
Revise 'input_check' to 'input_assure'; enforce JSON key alteration to match the sample ID if a mismatch is detected

### PR DESCRIPTION
**Pull request by kylacochrane**
This PR resolves an issue that arises when a `.mlst.json` file, generated by `Locidex`, retains the original IRIDA Next sample identifier after a sample is cloned into a new project, leading to a mismatch in identifiers.

To address this, the PR alters the previous `input_check` process, which now reads the `.mlst.json` file from `Locidex` and if the sample identifier does not match the JSON key, it is overwritten to ensure consistency. 

The process has been renamed to `input_assure` for clarity.

An `error_report.csv` is generated to identify any samples where the JSON key has been forcefully altered and discloses whether they are a query or reference sample in the pipeline. 

Previous tests still need to be updated. An additional test where the JSON file is gzipped will also be added.

### CODE REVIEW

File: workflows/gas_nomenclature.nf [hunk 1]
```diff
@@ -72,29 +72,18 @@ workflow GAS_NOMENCLATURE {
...
-        if (meta.id_match) {
-            return true // Keep the sample
-        } else if (meta.address == null && !meta.id_match) {
-            return false // Remove the sample
-        } else if (meta.address != null && !meta.id_match) {
-            // Exit with error statement
-            throw new RuntimeException("Pipeline exiting: sample with ID ${meta.id} does not have matching MLST JSON file.")
-        }
-    }
+    match.view()
```
Comment (emarinier) [inline workflows/gas_nomenclature.nf:83]: Do we want to leave the `.view()` in or was this for debugging?

File: workflows/gas_nomenclature.nf [hunk 2]
```diff
@@ -72,29 +72,18 @@ workflow GAS_NOMENCLATURE {
...
 
     // Prepare reference and query TSV files for LOCIDEX_MERGE
-    profiles = new_input.branch{
+    profiles = match.branch {
         query: !it[0].address
     }
```
Comment (emarinier) [inline workflows/gas_nomenclature.nf:88]: Is this for only querying profiles that have don't already have an address?

**Review by emarinier — CHANGES_REQUESTED** [review]

File: workflows/gas_nomenclature.nf (hunk 1 above)
Comment (kylacochrane) [reply workflows/gas_nomenclature.nf]: Haha - we don't. 
Although highlighting this also made me realize that we no longer need to store the `id_match` boolean in the `meta` data as we won't be removing any samples from the analysis.

I have updated the python code, the workflow, and the `input_assure` process to simplify this here: [95e40f6](https://github.com/phac-nml/gasnomenclature/pull/13/commits/95e40f6979ab60cbe6f8531bb72996c45cf1c56f)

File: workflows/gas_nomenclature.nf (hunk 2 above)
Comment (kylacochrane) [reply workflows/gas_nomenclature.nf]: Yes exactly 😄

**Review by apetkau — CHANGES_REQUESTED** [review]
This is amazing @kylacochrane 😄 . Thanks so much.

In addition to the in-line comments, I'm wondering if you can also fix up the failing test.

File: bin/input_check.py [hunk 3]
```diff
@@ -4,35 +4,45 @@
 import argparse
 import sys
 import csv
+import gzip
 
+def open_file(file_path, mode):
```
[the remark this answers is not in the document]
Comment (kylacochrane) [reply bin/input_check.py]: 🤦 fixed here: [c7252cf](https://github.com/phac-nml/gasnomenclature/pull/13/commits/c7252cfeb77a4f32fce0f0811e711961ce07b19a)

**Comment by kylacochrane** [comment]
The following tests have been updated or added:

1. Added a test for mismatched JSON keys and sampleIDs, covering both reference and query samples. This single test replaces the two previous tests that checked for removal of mismatched samples from the pipeline.
[82c3a0d](https://github.com/phac-nml/gasnomenclature/pull/13/commits/82c3a0d9ceea539a45cfe5c8b57b488a4a4657cf)

2. Added tests for scenarios where the MLST JSON file contains multiple entries (keys). Two tests were added: one verifying a match between a key and the sampleID, and another where none of the multiple keys match the sampleID.
[8e8ffa4](https://github.com/phac-nml/gasnomenclature/pull/13/commits/8e8ffa446b816d973e5aa04888cc26d75b9c18b4)
[7c1b5dc](https://github.com/phac-nml/gasnomenclature/pull/13/commits/7c1b5dc31536243a5870ed050f5b90286b33ff67)

3. Introduced a test for handling gzipped MLST JSON files to ensure input_assure can manage compressed files effectively.
[3266330](https://github.com/phac-nml/gasnomenclature/pull/13/commits/32663302aa75ba71c3b606c08b2ec62e3dea3c03)

File: bin/input_assure.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    # Write file containing relevant error messages
+    if error_message:
+        with open(output_error_file, "w", newline="") as f:
+            writer = csv.writer(f)
+            writer.writerow(["sample", "JSON_key", "error_message"])
+            writer.writerow([sample_id, keys, error_message])
+
+    # Write the updated JSON data back to the original file
+    with open_file(json_file, "wt") as f:
```
Comment (apetkau) [inline bin/input_assure.py:58]: Could the code be changed to write the updated JSON to a new file?

In Nextflow, sometimes inputs are passed as symlinks, and so overwriting the original file may lead to overwriting the input files to a pipeline. I also know nextflow excludes matches to inputs in the work directory in certain cases, but not sure what triggers this.

**Review by apetkau — CHANGES_REQUESTED** [review]
Thanks so much Kyla for all your hard work and the updates you've made 😄 

I have just one additional comment.

File: bin/input_assure.py (hunk 4 above)
Comment (kylacochrane) [reply bin/input_assure.py]: Ahh ok! Good to know 😃 
Change made here to ensure a new JSON file is produced by `input_assure`: [6642b72](https://github.com/phac-nml/gasnomenclature/pull/13/commits/6642b72ad0132805bf23083cf9979a8dde965941)

Comment (kylacochrane) [reply bin/input_assure.py]: Oh! For sure - good to know:
Changes made here:[6642b72](https://github.com/phac-nml/gasnomenclature/pull/13/commits/6642b72ad0132805bf23083cf9979a8dde965941)

File: modules/local/input_assure/main.nf [hunk 5]
```diff
@@ -11,19 +10,19 @@ process INPUT_CHECK{
...
 
     """
-    input_check.py \\
+    input_assure.py \\
         --input ${mlst} \\
         --sample_id ${meta.id} \\
         --address ${meta.address} \\
         --output_error ${meta.id}_error_report.csv \\
-        --output_match ${meta.id}_match.txt
+        --output_json ${meta.id}.mlst.json
```
Comment (apetkau) [inline modules/local/input_assure/main.nf:25]: Could you change to output gzipped JSON.

**Review by apetkau — CHANGES_REQUESTED** [review]
Thanks so much for all the changes Kyla 😄 . Just one more comment.

File: modules/local/input_assure/main.nf (hunk 5 above)
Comment (kylacochrane) [reply modules/local/input_assure/main.nf]: Done here 😸  : [348fe95](https://github.com/phac-nml/gasnomenclature/pull/13/commits/348fe9558c27e1ef76ba387c4abc31632656c1aa)

**Review by apetkau — APPROVED** [review]
This is amazing @kylacochrane . Thanks so much for all your hard work 😄

**Pull request merged**